Takes View Model- JVM

KJoslyn opened the pull request:
**What is included:**
- Takes tabpane is now populated with the selected resource's content and takes
- "New Take" button opens the preferred audio plugin and creates a new take for the selected resource
- Only one take can be played at a time. If a take is playing and another take is played, the first one will pause.

**What is not included:**
- Editing, deleting, selecting takes
- Properly formatted markdown (what you see is the raw markdown content)
- Navigation (you can navigate between the resources list and takes page by checking out the kj-takes-testapp branch, but this code is not in the PR.)

**How to test:**
1. Check out `kj-takes-testapp` branch (based on this PR's branch), and run `main` in ResourceTakesApp.kt. Also make sure to checkout the corresponding branch in common, `kj-takes-tabpane-newtake` (sister PR [#105](https://github.com/WycliffeAssociates/otter-common/pull/105) in common)
2. A chapter will be loaded by default, after which you can select "View Recordings" to go to the takes page for any resource. 
3. Try creating new takes
4. Play a take, and try playing a take while another take is playing
5. Switch between the "snippet" and "note" tabs
6. Return to the resource list page by clicking the "back" button in the header. (This header as well as the file menu are not in this PR, they are only in the kj-takes-testapp branch.)

KJoslyn committed:
Resource takes page view

KJoslyn committed:
Test app for resource take cards

KJoslyn committed:
Added test wav files

KJoslyn committed:
Renamed test app

KJoslyn committed:
Navigate to takes from resources list (load real data)

KJoslyn committed:
Takes view model

KJoslyn committed:
Load test text and takes

KJoslyn committed:
TakesListView and Play/Pause event handling

KJoslyn committed:
Takes page refactor

Fix broken Takes Fragment

KJoslyn committed:
Support change to Resource

KJoslyn committed:
Refactoring to support RecordableItem

KJoslyn committed:
Use observable list of Recordables rather than 'title' and 'body'

KJoslyn committed:
Better names for Takes fragments

KJoslyn committed:
WorkbookViewModel

KJoslyn committed:
navigateToTakesPage() uses BookElement

KJoslyn committed:
Support change to RecordTake

KJoslyn committed:
New take action

KJoslyn committed:
Put tab label properties in an enum map

KJoslyn committed:
Add sort to tabs

KJoslyn committed:
Fixed bug in TakeDao

KJoslyn committed:
Proper subscribe handling

KJoslyn committed:
Add vgrow to workspace root to fill screen

KJoslyn committed:
Dispose takes subscription when Recordable changes

KJoslyn committed:
Remove test files to be put in test branch

KJoslyn committed:
WorkbookViewModel getters return non-null values or throw error

KJoslyn committed:
Remove code dependent on testapp files

KJoslyn committed:
Utility function for EnumMap: getNotNull()

KJoslyn committed:
Bug fix for TakesTab clear disposables

KJoslyn committed:
Restore functionality to TMVM recordContent (sort of)

TakeManagementViewModel will be merged with TakesViewModel soon, so these changes will be obsolete when that happens.

KJoslyn commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/resourcetakes/viewmodel/TakesViewModel.kt:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    fun setRecordableListItems(items: List<Recordable>) {
+        if (!recordableList.containsAll(items))
```
This makes use of the overriden equals() method on Recordable

KJoslyn commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/workbook/viewmodel/WorkbookViewModel.kt:
```diff
@@ -0,0 +1,36 @@
+class WorkbookViewModel: ViewModel() {
+    val activeWorkbookProperty = SimpleObjectProperty<Workbook?>()
+    val workbook: Workbook
+        get() = activeWorkbookProperty.value ?: throw IllegalStateException("Workbook is null")
```
Rather than performing null (or ?.let) checks everywhere and dirtying up the code, I elected to make these getters throw errors if the value is null.
Since all of the getters in WorkbookViewModel follow this pattern, there should be no surprises as long as one knows this is the expected behavior.

KJoslyn replied:
Update 6/26: Unfortunately I don't think the pattern of throwing an exception on null can apply to every property since Chunk is actually nullable (sometimes we are recording material about a Chapter, not a particular Chunk.)

KJoslyn commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/takemanagement/viewmodel/TakeManagementViewModel.kt:
```diff
@@ -31,6 +32,9 @@ class TakeManagementViewModel : ViewModel() {
     private val pluginRepository = injector.pluginRepository
 
+    private val takesViewModel: TakesViewModel by inject()
+    private val workbookViewModel: WorkbookViewModel by inject()
```
These are added so that we can call the new record() function (line 160.) However, TakeManagementViewModel and TakesViewModel will be merged soon, so the changes to this file will be obsolete when that happens.

KJoslyn replied:
Update 6/26: I removed takesViewModel from here. TakesViewModel actually got renamed to RecordResourceViewModel, which does not contain any recording business logic. TakeManagementViewModel is now the thing that houses logic for recording/editing/etc Takes (and a lot more). 

Likewise, I think we should have a RecordScriptureViewModel that only contains things needed to bind data to the "scripture takes" view. The current TakeManagementViewModel probably needs to be greatly refactored.

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/resourcetakes/view/ResourceTakesStyles.kt:
```diff
@@ -0,0 +1,91 @@
+        }
+
+        leftRegionContainer {
+            borderColor += box(Color.TRANSPARENT, AppTheme.colors.lightBackground, AppTheme.colors.lightBackground, Color.TRANSPARENT)
```
Line is a little long.

KJoslyn replied:
True, fixing that.

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/resourcetakes/view/ResourceTakesView.kt:
```diff
@@ -0,0 +1,70 @@
+    private val viewModel: TakesViewModel by inject()
+    private val tabPane = ChromeableTabPane()
+
+    class ContentTypeToTabMap(map: Map<ContentType, TakesTab?>): EnumMap<ContentType, TakesTab>(map)
```
Does the value need to be nullable?

KJoslyn replied:
Oops, not anymore. It was leftover from before I made the `getNotNull()` function, when nulls were propagating too far.

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/resourcetakes/view/ResourceTakesView.kt:
```diff
@@ -0,0 +1,70 @@
+    private val tabPane = ChromeableTabPane()
+
+    class ContentTypeToTabMap(map: Map<ContentType, TakesTab?>): EnumMap<ContentType, TakesTab>(map)
+    private val contentTypeToTabMap = ContentTypeToTabMap(
```
Subclassing doesn't seem necessary. Can't contentTypeToTabMap just be an EnumMap?

KJoslyn replied:
Actually sorry I meant to comment on this. In MapUtils.kt, `getNotNull()` uses reflection to get the class name when throwing an exception. That's really the only reason I subclassed EnumMap.

aunger replied:
I might consider just making getNotNull() a method of ContentTypeToTabMap. I'm ok with this as-is, though.

jsarabia replied:
can we avoid nulls altogether here?

KJoslyn replied:
Not quite, we can't guarantee that a key exists in the map

Commit:
PR Comments

jsarabia commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/resources/viewmodel/ResourcesViewModel.kt:
```diff
@@ -1,43 +1,48 @@
+    private val takesViewModel: TakesViewModel by inject()
+    private val workbookViewModel: WorkbookViewModel by inject()
 
     val resourceGroups: ResourceGroupCardItemList = ResourceGroupCardItemList(mutableListOf())
```
If this is a usual use case, perhaps the initialization of an empty mutable list could be done by default and thus not have a parameter here?

KJoslyn replied:
Sure, good point

jsarabia commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/resourcetakes/view/ResourceTakesStyles.kt:
```diff
@@ -0,0 +1,96 @@
+    init {
+        main {
+            prefWidth = Screen.getPrimary().visualBounds.width.px - 600.0
+            prefHeight = Screen.getPrimary().visualBounds.height.px - 200.0
```
Seems like a bad idea, and as a placeholder it will likely be forgotten to be fixed?

KJoslyn replied:
Yeah I don't think this shouldn't be here anymore, I'll remove it

KJoslyn committed:
PR Comments

jsarabia commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/resourcetakes/view/TabContentLeftRegion.kt:
```diff
@@ -0,0 +1,49 @@
+        vgrow = Priority.ALWAYS
+        vbox {
+            addClass(ResourceTakesStyles.dragTarget)
+            text("Drag Take Here")
```
messages

jsarabia commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/resourcetakes/viewmodel/TakesViewModel.kt:
```diff
@@ -0,0 +1,82 @@
+        LaunchPlugin(pluginRepository)
+    )
+
+    class ContentTypeToLabelPropertyMap(map: Map<ContentType, SimpleStringProperty?>):
```
Why nullable?

KJoslyn replied:
Good catch, it shouldn't be nullable

jsarabia commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/resources/viewmodel/ResourcesViewModel.kt:
```diff
@@ -1,43 +1,48 @@
-            .subscribe {
-                Platform.runLater {
-                    resourceGroups.addAll(it)
+        workbookViewModel.chapter.let { chapter ->
```
why use let here?

KJoslyn replied:
I use `chapter` again at line 23 so I didn't want to write `workbookViewModel.chapter` again, but maybe I should have just done `val chapter = workbookViewModel.chapter` at the start

jsarabia commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/resources/viewmodel/ResourcesViewModel.kt:
```diff
@@ -1,43 +1,48 @@
+                .subscribe {
+                    Platform.runLater {
+                        resourceGroups.addAll(it)
+                    }
```
This should probably happen in the view controller (view/fragment)

KJoslyn replied:
Isn't the observable list usually contained and manipulated in the view model, and the view/fragment simply listens for changes to the list?

KJoslyn replied:
So it turns out that we don't need to use Platform.runLater when we are adding items to an observable list. The listview will automatically notice changes to the observable list and add the items anyway. I was confused because in the case where resource cards are added to the resource group card, we are actually directly adding a node to the node graph- in which case we need Platform.runLater.

By removing Platform.runLater, we're closer to being able to move this ViewModel into common!

Edit: Unfortunately I'm still a little confused because it seems that I need the Platform.runLater in the loadTakes() function of RecordableTabViewModel. If it isn't there, the listview does not update correctly- I get a duplicate take card showing up when I record a take. I want to understand this!

jsarabia commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/resourcetakes/view/TakesTab.kt:
```diff
@@ -0,0 +1,106 @@
+        disposables.clear()
+    }
+
+    private fun getFormattedTextBinding() = Bindings.createStringBinding(Callable { getFormattedText() }, recordableProperty)
```
line length

KJoslyn committed:
PR Comments and applying FileName pattern

KJoslyn committed:
Using Recordable Interface and WorkbookFileNamerBuilder

KJoslyn committed:
Simplified WorkbookFileNameBuilder

KJoslyn committed:
Using simplified FileNamer

KJoslyn committed:
Better filenames

KJoslyn commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/workbook/viewmodel/WorkbookViewModel.kt:
```diff
@@ -0,0 +1,35 @@
+        get() = activeChapterProperty.value ?: throw IllegalStateException("Chapter is null")
+
+    val activeChunkProperty = SimpleObjectProperty<Chunk?>()
+    val chunk by activeChunkProperty
```
See above comment about getters in this file

KJoslyn committed:
Renamed to RecordResourceViewModel

Moved recording functionality to TakeManagementViewModel (which needs to be refactored later)

KJoslyn committed:
Removed Platform.runLater where it is not necessary

KJoslyn committed:
Refactored RecordableTab, added RecordableTabViewModel

KJoslyn committed:
Minor improvements

KJoslyn committed:
ResourcesViewModel tests and RecordableTab tests

jsarabia commented:
how can we update travis to get the chromable tab pane so the tests work?

aunger committed:
Allow importing OBS markdown files (#461)

* Allow importing OBS markdown files.

MimeType enum instead of string literals.
Ignore creator if unable to match help RCs otherwise.
Unhide the Tree generics.
More type safety.

* Narrow down a few more OtterTree<Any> params.

KJoslyn committed:
Test cleanup

Commit:
Merge branch 'dev' into kj-takes-tabpane-newtake

KJoslyn committed:
Fix broken unit tests

KJoslyn committed:
Update travis.yml to include chromeabletabpane

KJoslyn committed:
Remove RecordableTabTest until better solution found

It used PlatformImpl.startup, which errors on travis
